processing/src/main/java/org/apache/druid/segment/filter/OrFilter.java

```diff
@@ -82,158 +84,151 @@ public <T> FilterBundle makeFilterBundle(
+              index = bundle.getIndex().getBitmap();
+            } else {
+              index = index.union(bundle.getIndex().getBitmap());
+            }
+          } else {
+            // index and matcher bundles must be handled separately, they will need to be a single value matcher built
+            // by doing an AND operation between the index and the value matcher
+            // (a bundle is basically an AND operation between the index and matcher if the matcher is present)
+            partialIndexBundles.add(convertBundleToMatcherOnlyBundle(bundle, bundleIndex));
           }
-          indexFilters.add(subfilter);
-          needMatcher = !index.getIndexCapabilities().isExact();
-        } else if (!allowPartialIndex) {
-          break;
         }
-      } else if (!allowPartialIndex) {
-        break;
-      }
-
-      if (needMatcher) {
-        matcherFilters.add(subfilter);
+      } else {
+        matcherOnlyBundles.add(bundle.getMatcherBundle());
       }
     }
     final long totalBitmapConstructTimeNs = System.nanoTime() - bitmapConstructionStartNs;
 
-    final ImmutableBitmap index;
-    final ImmutableBitmap partialIndex;
-    if (indexFilters.size() == filters.size()) {
-      // all or nothing
-      index = bitmapResultFactory.toImmutableBitmap(bitmapResultFactory.union(indexes));
-      partialIndex = null;
-    } else {
-      index = null;
-      if (allowPartialIndex) {
-        if (indexes.isEmpty()) {
-          partialIndex = null;
-        } else if (indexes.size() == 1) {
-          partialIndex = bitmapResultFactory.toImmutableBitmap(indexes.get(0));
-        } else {
-          partialIndex = bitmapResultFactory.toImmutableBitmap(bitmapResultFactory.union(indexes));
-        }
-      } else {
-        partialIndex = null;
-        matcherFilters.clear();
-        // need to convert to all matchers
-        matcherFilters.addAll(filters);
+
+    // if all the filters are 'index only', we can make an index only bundle
+    if (indexOnlyBundles.size() + emptyCount == filters.size()) {
+      if (index == null || index.isEmpty()) {
+        return FilterBundle.allFalse(
+            totalBitmapConstructTimeNs,
+            columnIndexSelector.getBitmapFactory().makeEmptyImmutableBitmap()
+        );
+      }
+      if (indexOnlyBundles.size() == 1) {
+        return new FilterBundle(
+            new FilterBundle.SimpleIndexBundle(
+                indexOnlyBundles.get(0).getIndexInfo(),
+                index,
+                indexOnlyBundles.get(0).getIndexCapabilities()
+            ),
+            null
+        );
       }
+      return new FilterBundle(
+          new FilterBundle.SimpleIndexBundle(
+              new FilterBundle.IndexBundleInfo(
+                  () -> "OR",
+                  selectionRowCount,
+                  totalBitmapConstructTimeNs,
+                  indexOnlyBundles.stream().map(FilterBundle.IndexBundle::getIndexInfo).collect(Collectors.toList())
```

Comment:
I'm not sure how often this is going to be called, but I've found that the performance of the java streaming API is pretty abysmal.  If this was forced into a lazy evaluation that only matter when we are doing debug logging, I wouldn't care, but it looks like it's forced CPU and usage here...

Is there a way to shove this into the lambda for the toString instead (so that this isn't done unless it's actually necessary?)

Reply:
it should only be called once, but I can make it an iterable on the bundles so that can make this lazy the same way the string is